clients/p4rt-ctl-cpp/p4rt_perf_session.cc

```diff
@@ -0,0 +1,184 @@
+  }
+
+  return std::move(session);
+}
+
+// Creates a session with the switch, which lasts until the session object is
+// destructed.
+absl::StatusOr<std::unique_ptr<P4rtSession>> P4rtSession::Create(
+    const std::string& address,
+    const std::shared_ptr<grpc::ChannelCredentials>& credentials,
+    uint32_t device_id, absl::uint128 election_id) {
+  /* OVS spawns multiple revalidator threads and handler threads to handle
+   * datapath Notifications.
+   * When multiple L2 MAC's are learnt at the same time, the learn information
+   * can come from revlidator thread or handler thread. On a worst case
+   * scenario, both the threads will call P4rtSession::Create exactly at
+   * the same time and both threads can get same election_id based on API
+   * TimeBasedElectionId(). This causes one of the threads to get stuck and
+   * no further data path information is received by the thread.
+   * To avoid this issue, we add Thread ID to the election ID to generate an
+   * unique election_id, even when two threads receive MAC learn at the same
+   * time.
+   */
+  election_id = election_id + (absl::uint128)pthread_self();
+  return Create(CreateP4RuntimeStub(address, credentials), device_id,
+                election_id);
+}
+
+absl::Status GetForwardingPipelineConfig(P4rtSession* session,
+                                         p4::config::v1::P4Info* p4info) {
+  GetForwardingPipelineConfigRequest request;
+  request.set_device_id(session->DeviceId());
+  request.set_response_type(
+      GetForwardingPipelineConfigRequest::P4INFO_AND_COOKIE);
+
+  GetForwardingPipelineConfigResponse response;
+  grpc::ClientContext context;
+  absl::Status status =
+      GrpcStatusToAbslStatus(session->Stub().GetForwardingPipelineConfig(
+          &context, request, &response));
+
+  *p4info = response.config().p4info();
+
+  return absl::OkStatus();
+}
+
+absl::StatusOr<ReadResponse> SendReadRequest(P4rtSession* session,
+                                             const ReadRequest& read_request) {
+  grpc::ClientContext context;
+  auto reader = session->Stub().Read(&context, read_request);
+
+  ReadResponse response;
+  ReadResponse partial_response;
+  while (reader->Read(&partial_response)) {
+    response.MergeFrom(partial_response);
+  }
+
+  grpc::Status reader_status = reader->Finish();
+  if (!reader_status.ok()) {
+    return GrpcStatusToAbslStatus(reader_status);
+  }
+
+  return std::move(response);
+}
+
+absl::Status SendWriteRequest(P4rtSession* session,
+                              const WriteRequest& write_request) {
+  grpc::ClientContext context;
+  WriteResponse response;
+
+  ::grpc::Status status =
+      session->Stub().Write(&context, write_request, &response);
+
+  return GrpcStatusToAbslStatus(status);
+}
+
+::p4::v1::TableEntry* SetupTableEntryToInsert(P4rtSession* session,
+                                              ::p4::v1::WriteRequest* req) {
+  req->set_device_id(session->DeviceId());
```

Comment:
Only Update is of type repeated in this WriteRequest message. Before calling this API in loop, we can set device_id, election_id only once for any writerequest ?

snippet:

message WriteRequest {
  uint64 device_id = 1;
  uint64 role_id = 2 [deprecated=true];
  string role = 6;
  Uint128 election_id = 3;
  // The write batch, comprising a list of Update operations. The P4Runtime
  // server may arbitrarily reorder messages within a batch to maximize
  // performance.
  repeated Update updates = 4;

Reply:
Done.